File: src/json_repair.ts
export class JsonExtractionError extends Error {
  constructor(
    message: string,
    public readonly rawText: string,
  ) {
    super(message);
    this.name = "JsonExtractionError";
  }
}

function stripCodeFences(text: string): string {
  const fenced = text.match(/```(?:json)?\s*\n?([\s\S]*?)```/);
  return fenced ? fenced[1].trim() : text;
}

function extractBracketedJson(text: string): string {
  const first = text.indexOf("{");
  const last = text.lastIndexOf("}");
  if (first === -1 || last === -1 || last <= first) {
    throw new JsonExtractionError("No JSON object delimiters found", text.slice(0, 200));
  }
  return text.slice(first, last + 1);
}

function repairCommonIssues(text: string): string {
  let repaired = text.replace(/,\s*([}\]])/g, "$1");

  repaired = repaired.replace(/(?<=[:,[{]\s*)'((?:[^'\\]|\\.)*?)'\s*(?=[,\]}\n])/g, '"$1"');
  repaired = repaired.replace(/(?<=\{|,)\s*'((?:[^'\\]|\\.)*?)'\s*:/g, ' "$1":');

  return repaired;
}

export function extractJson<T>(text: string): T {
  const stripped = stripCodeFences(text);
  const candidate = extractBracketedJson(stripped);

  try {
    return JSON.parse(candidate) as T;
  } catch {
    // fall through to repair attempt
  }

  const repaired = repairCommonIssues(candidate);
  try {
    return JSON.parse(repaired) as T;
  } catch (e) {
    throw new JsonExtractionError(
      `Failed to parse JSON after repair: ${e instanceof Error ? e.message : e}`,
      candidate.slice(0, 200),
    );
  }
}

File: python/json_repair.py
from __future__ import annotations

import json
import re


class JSONExtractionError(Exception):
    pass


def extract_json(text: str) -> dict:
    cleaned = _strip_code_fences(text)
    raw = _extract_braces(cleaned)
    return _parse_with_repairs(raw, text)


def _strip_code_fences(text: str) -> str:
    return re.sub(r"```(?:json)?\s*\n?([\s\S]*?)\n?\s*```", r"\1", text)


def _extract_braces(text: str) -> str:
    first = text.find("{")
    last = text.rfind("}")
    if first == -1 or last == -1 or last <= first:
        raise JSONExtractionError(
            f"No JSON object found in text: {text[:120]!r}"
        )
    return text[first : last + 1]


def _parse_with_repairs(raw: str, original_text: str) -> dict:
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        pass

    repaired = _fix_trailing_commas(raw)
    try:
        return json.loads(repaired)
    except json.JSONDecodeError:
        pass

    repaired = _fix_single_quotes(repaired)
    try:
        return json.loads(repaired)
    except json.JSONDecodeError as exc:
        raise JSONExtractionError(
            f"Failed to parse JSON after repairs: {exc}. "
            f"Original text: {original_text[:200]!r}"
        ) from exc


def _fix_trailing_commas(text: str) -> str:
    return re.sub(r",\s*([}\]])", r"\1", text)


def _fix_single_quotes(text: str) -> str:
    return re.sub(r"(?<![\\])'", '"', text)
